README.md
# Framer 

[![build status](https://secure.travis-ci.org/mbrevoort/framer.png)](http://travis-ci.org/mbrevoort/framer)

Framer is a simple, dynamic file proxy and photo resizing http server intended to be behind an http cache or act as a CDN root server. Framer accepts uploads and stores it's files in Amazon S3.

Framer does not cache anything itself but rather sets a max-age Cache-Control header so the browser or caching proxy should act appropriately. It keeps the files it manages in S3 side-stepping the burden of managing, syncronizing and backing files across instances essentially making the instances stateless.

Lastly, Framer will generate a path to each file prefixed with a `uuid` that should be difficult to guess. All files are essentially public and accessible if you know the URI.

## Usage

Let's cut to the chase. Look in the example directory at `server.js` for a working example.

### Installing

```
npm install framer
```

# Uploading

Framer accepts multipart uploads. It takes the raw file an streams it into your S3 bucket.

```
// configure, passing an optional authHandler function to handle upload 
// authorization. If ommitting, any client can upload unauthenticated
var handleUpload = framer.handleUpload({
  authHandler: function (authValue, cb) {
    // check if authorization value is valid, if so execute callback 
    // without an error and pass along a user identifier to prefix
    // the path of the location to store the image so that you know
    // who owns which images and perhaps authorization can be applied
    // this way in the future
    var userId = userIdFromToken(authValue);
    if (userId)
      cb(null, userId);
    else
      cb(new Error('unauthorized));
  },
  prefix: '/img'  // uri prefix you expect to serve images from
});

if (req.url === '/upload' && req.headers['content-type']) {
  handleUpload(req, res);
}
```

The result of a successful upload will be an `application/json` response with a `uri` property containing the relative URI of the new resource. *Note*: this doesn't include the protocol, host or port and will be prefixed with `opts.prefix` if specified. For example:

```
{
    "statusCode": 200,
    "uri": "/prefix/raw/c84d9b70-caca-11e2-8e30-ab79663612ac/image.jpg"
}
```

If the upload fails, the respons will look something like this:

```
{
    "statusCode": 403,
    "error": "some ugly error message"
}
```

# Serving

Once uploaded, the images can be served from the result URL. Look at the example
to more easily understand how to use it. Basically you would wire it up like this:

```
// configure
var serveImage = framer.serveImage({ prefix: '/img', cacheMaxAge: 3600 });
var serveFile = framer.serveFile({ prefix: '/file', cacheMaxAge: 3600 });

// handle a request: 
if (req.url.indexOf('/img/') === 0) serveImage(req, res);
else if (req.url.indexOf('/file/') === 0) serveImage(req, res);
```

## Serving Images

Framer provides a simple API for resizing images on the fly to suite your needs.
This is helpful when you need multiple sizes or an image, and it's not practical
to generate all of the sizes up front.

The most useful is scale, center and crop. For example, let's say you have need
to render 100x100 thumbnails of images but that have different dimensions and size.
To scale a 1200x800 image to be a 100x100 square without changing the aspect ratio,
we do this in two stages. First we scale the height to 100 pixels resulting in an
intermediate image that's now 150x100, then we orient to the center of the image and
and crop the width to the center 100 pixel. The resulting image is 100x100, cropping out
25 pixels from each side or 200 pixels from each size of the original image.

There are two other options for scaling images that will either result in a result that
will fit into the wxh box or fill the wxh box, resulting in images 100x67 and 150x100,
respectively if considering the example above.

Images are served by passing the request and response to the function produced by 
`serveImage`.

### Original Image

Use `raw`:

`/raw/<path_to_image>`

### Scale, Center, Crop

Use a `x` to delimit width and height:

`/100x100/<path_to_image>`

### Scale within Box

Use a `-` to delimit width and height:

`/100-100/<path_to_image>`

### Scale within Box

Use a `+` to delimit width and height:

`/100+100/<path_to_image>`


## Serving files

Other non image files may be uploaded and served as well. Files are served by passing 
the request and response to the function produced by `serveFile` without the sizing 
prefix of serveImage:

`<path_to_file>`


# Contributing

Pull request and issues are totally welcome. If you have quick questions hit me up on Twitter @mbrevoort.


CHANGELOG.md
### 0.2.0

* added `opts.prefix` option to `handleUpload` and renamed `uriSuffix` to `uri` in upload response
* added jshint to tests

### 0.1.0

* First version. Surely there will be more to come!

test/test.upload.js
var assert = require('assert')
  , fs = require('fs')
  , http = require('http')
  , path = require('path')
  , uuid = require('node-uuid')
  , request = require('request')
  , Framer = require('..')
  ;

describe('test upload', function () {

  var expectedFile = null;
  var s3Options = {
    secure: false,
    key: 'bogus',
    secret: 'bogus',
    bucket: 'bogus',
  };
  var s3Client = {
    putStream: function (part, destPath, headers, cb) {
      var response = {
        statusCode: 200
      };
      accumulateStream(part, function (err, result) {
        assert.ifError(err);
        assert(result);
        assert.equal(expectedFile.toString(), result.toString());
        cb(null, response);
      })
    }
  }

  before(function (done) {
    fs.readFile(path.join(__dirname, 'image.jpg'), function (err, data) {
      expectedFile = data;
      done();
    });
  })

  it('should take upload', function (done) {

    var framer = new Framer({
      s3: s3Options,
    });
    framer._s3Client = s3Client;
    var handleUpload = framer.handleUpload({ prefix: '/prefix' });

    var PORT = Math.ceil(Math.random()*2000 + 1024);
    var client = http.createServer(function (req, res) {
      handleUpload(req, res);
    }).listen(PORT);

    request
      .post('http://127.0.0.1:' + PORT + '/', { json: true }, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert(body.uri.indexOf('/prefix/') === 0, 'uri should be prefixed with the prefix option')
        done();
      })
      .form().append("filename", fs.createReadStream(path.join(__dirname, "image.jpg")));
  });

  it('should delegate authoritation', function (done) {

    var authValue = 'foobarbaz';
    var authHandlerCalled = false;
    var authHandler = function (value, cb) {
      authHandlerCalled = true;
      assert.equal(authValue, value);
      cb(null, value);
    }
    var framer = new Framer({
      s3: s3Options
    });
    framer._s3Client = s3Client;
    var handleUpload = framer.handleUpload({ authHandler: authHandler });

    var PORT = Math.ceil(Math.random()*2000 + 1024);
    var client = http.createServer(function (req, res) {
      handleUpload(req, res);
    }).listen(PORT);

    var r = request.post('http://127.0.0.1:' + PORT + '/', function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert(authHandlerCalled);
        done();
      })
    var f = r.form();
    f.append("authorization", authValue)
    f.append("filename", fs.createReadStream(path.join(__dirname, "image.jpg")));
  });


  it('should reject invalid authoritation', function (done) {

    var authValue = 'foobarbaz';
    var authHandlerCalled = false;
    var authHandler = function (value, cb) {
      authHandlerCalled = true;
      cb(new Error('unauthorized'));
    }
    var framer = new Framer({
      s3: s3Options
    });
    framer._s3Client = s3Client;
    var handleUpload = framer.handleUpload({ authHandler: authHandler });

    var PORT = Math.ceil(Math.random()*2000 + 1024);
    var client = http.createServer(function (req, res) {
      handleUpload(req, res);
    }).listen(PORT);

    var r = request.post('http://127.0.0.1:' + PORT + '/', function (err, res, body) {
        assert.ifError(err);
        assert.equal(401, res.statusCode);
        assert(authHandlerCalled);
        done();
      })
    var f = r.form();
    f.append("authorization", authValue)
    f.append("filename", fs.createReadStream(path.join(__dirname, "image.jpg")));
  });

});


function accumulateStream (stream, cb) {
  var bufs = [];
  stream.on('data', function (d){ bufs.push(d); });
  stream.on('error', cb);
  stream.on('end', function (){
    var buf = Buffer.concat(bufs);
    cb(null, buf);
  })
}

test/test.file-download.js
var assert = require('assert')
  , fs = require('fs')
  , http = require('http')
  , path = require('path')
  , uuid = require('node-uuid')
  , request = require('request')
  , Framer = require('..')
  ;

describe('test image downloads', function () {
  var prefix = 'prefix',
      uid = uuid.v1(),
      maxAge = 3600,
      expectedFile = null;

  var s3Options = {
    secure: false,
    key: 'bogus',
    secret: 'bogus',
    bucket: 'bogus'
  };

  var s3Client = {
    get: function (imagePath) {
      var expectedPath = '/' + prefix + '/' + uid + '/file.txt';

      var mock = {};
      mock.on = function (evt, cb) {
        if (evt === 'response') {
          if (imagePath === expectedPath) {
            var obj = fs.createReadStream(path.join(__dirname, 'file.txt'));
            obj.statusCode = 200;
            cb(obj);
          }
          else {
            cb({ statusCode: 403, on: function (){} });
          }
        }
        return mock;
      };
      mock.end = function () {};
      return mock;
    }
  };

  var PORT = Math.ceil(Math.random()*2000 + 1024);
  var framer = new Framer({
    s3: s3Options
  });
  framer._s3Client = s3Client;

  var serveFile = framer.serveFile({ prefix: '/file', cacheMaxAge: maxAge });
  var client = http.createServer(function (req, res) {
    serveFile(req, res);
  }).listen(PORT);

  before(function (done) {
    client.listen(PORT, done);
  });

  it('should download non image file', function (done) {
    fs.readFile(path.join(__dirname, 'file.txt'), function (err, data) {
      var expectedFile = data;
      var url = 'http://127.0.0.1:' + PORT + '/file/' + prefix + '/' + uid + '/file.txt';

      request(url, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert.equal('max-age: ' + maxAge, res.headers['cache-control']);
        assert.equal(expectedFile.toString(), body.toString());
        done();
      });
    });
  });

});


test/test.image-download.js
var assert = require('assert')
  , fs = require('fs')
  , http = require('http')
  , path = require('path')
  , uuid = require('node-uuid')
  , gm = require('gm')
  , request = require('request')
  , buffertools = require('buffertools')
  , Framer = require('..')
  ;

describe('test image downloads', function () {
  var prefix = 'prefix',
      uid = uuid.v1(),
      maxAge = 3600,
      expectedFile = null;

  var s3Options = {
    secure: false,
    key: 'bogus',
    secret: 'bogus',
    bucket: 'bogus'
  };

  var s3Client = {
    get: function (imagePath) {
      var expectedPath = '/' + prefix + '/' + uid + '/image.jpg';

      var mock = {};
      mock.on = function (evt, cb) {
        if (evt === 'response') {
          if (imagePath === expectedPath) {
            var obj = fs.createReadStream(path.join(__dirname, 'image.jpg'));
            obj.statusCode = 200;
            cb(obj);
          }
          else {
            cb({ statusCode: 403, on: function (){} });
          }
        }
        return mock;
      };
      mock.end = function () {};
      return mock;
    }
  };

  var PORT = Math.ceil(Math.random()*2000 + 1024);
  var framer = new Framer({
    s3: s3Options
  });
  framer._s3Client = s3Client;

  var serveImage = framer.serveImage({ prefix: '/img', cacheMaxAge: maxAge });
  var client = http.createServer(function (req, res) {
    serveImage(req, res);
  }).listen(PORT);

  before(function (done) {
    client.listen(PORT, done);
  });

  it('should download raw image', function (done) {
    fs.readFile(path.join(__dirname, 'image.jpg'), function (err, data) {
      var expectedFile = data;
      var url = 'http://127.0.0.1:' + PORT + '/img/raw/' + prefix + '/' + uid + '/image.jpg';

      request(url, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert.equal(expectedFile.toString(), body.toString());
        done();
      });
    });
  });

  it('should resize crop center image', function (done) {
    var w=50, h=50;
    gm(fs.createReadStream(path.join(__dirname, 'image.jpg')))
    .resize(w, h, '^')
    .gravity('Center')
    .crop(w, h, 0, 0)
    .toBuffer(function (err, data) {
      var expectedFile = data;
      var url = 'http://127.0.0.1:' + PORT + '/img/50x50/' + prefix + '/' + uid + '/image.jpg';

      request(url, { encoding: null }, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert(expectedFile.equals(body))
        done();
      });
    });
  });

  it('should resize within box', function (done) {
    var w=50, h=50;
    gm(fs.createReadStream(path.join(__dirname, 'image.jpg')))
    .resize(w, h)
    .toBuffer(function (err, data) {
      var expectedFile = data;
      var url = 'http://127.0.0.1:' + PORT + '/img/50-50/' + prefix + '/' + uid + '/image.jpg';

      request(url, { encoding: null }, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert(expectedFile.equals(body))
        done();
      });
    });
  });

  it('should resize but fill box', function (done) {
    var w=50, h=50;
    gm(fs.createReadStream(path.join(__dirname, 'image.jpg')))
    .resize(w, h, '^')
    .toBuffer(function (err, data) {
      var expectedFile = data;
      var url = 'http://127.0.0.1:' + PORT + '/img/50+50/' + prefix + '/' + uid + '/image.jpg';

      request(url, { encoding: null }, function (err, res, body) {
        assert.ifError(err);
        assert.equal(200, res.statusCode);
        assert(expectedFile.equals(body))
        done();
      });
    });
  });

  it('should return max-age header', function (done) {
    var url = 'http://127.0.0.1:' + PORT + '/img/50+50/' + prefix + '/' + uid + '/image.jpg';

    request(url, { encoding: null }, function (err, res, body) {
      assert.ifError(err);
      assert.equal(200, res.statusCode);
      assert.equal('max-age: ' + maxAge, res.headers['cache-control']);
      done();
    });
  });

  it('should return 404 for unknown iamges', function (done) {
    var url = 'http://127.0.0.1:' + PORT + '/img/50+50/xxx/xxx/image.jpg';

    request(url, function (err, res, body) {
      assert.ifError(err);
      assert.equal(403, res.statusCode, 's3 returns 403 for unknown files');
      done();
    });
  });

});


index.js
var assert = require('assert')
  , multiparty = require('multiparty')
  , knox = require('knox')
  , uuid = require('node-uuid')
  , mime = require('mime')
  , Batch = require('batch')
  , gm = require('gm')
  ;

var Framer = module.exports = function Framer(opts) {
  if (typeof opts !== 'object') opts = {};

  var self = this;

  assert(typeof opts.s3 === 'object', 'opts.s3 required');
  assert(typeof opts.s3.key === 'string', 'opts.s3.key required');
  assert(typeof opts.s3.secret === 'string', 'opts.s3.secret required');
  assert(typeof opts.s3.bucket === 'string', 'opts.s3.bucket required');
  opts.s3.secure = (opts.s3.secure) ? true : false;

  this._s3Client = knox.createClient(opts.s3);

  this.handleUpload = function (opts) {
    if (!opts) opts = {};
    opts.prefix = opts.prefix || "";

    var makePublicReadable = opts.makePublicReadable || false
      , log = opts.log || function (){}
      , prefix = opts.prefix || ""
      , authHandler = opts.authHandler;

    return function (req, res) {
      var headers = {}
        , form = new multiparty.Form()
        , batch = new Batch();

      if (makePublicReadable) headers['x-amz-acl'] = 'public-read';

      var onUnexpectedEnd = function () {
        res.writeHead(500, {'content-type': 'text/plain'});
        res.end('incomplete upload');
      };

      batch.push(function(cb) {
        if (authHandler) {
          form.on('field', function(name, value) {
            if ( name === 'authorization') {
              authHandler(value, cb);
            }
          });
        }
        else cb(null, '');
      });

      batch.push(function(cb) {
        form.on('part', function(part) {
          if (! part.filename) return;
          cb(null, part);
        });
      });

      batch.end(function(err, results) {
        if (err) {
          return self._handleError(401, res, err);
        }

        form.removeListener('close', onUnexpectedEnd);

        var userPrefix = results[0]
          , part = results[1]
          , filename = part.filename
          , destPrefix = (userPrefix) ? '/' + userPrefix + '/' : '/'
          , destPath = encodeURI(destPrefix + uuid.v1() + '/' + filename)
          ;

        headers['Content-Length'] = part.byteCount;
        headers['Content-Type'] = mime.lookup(destPath);

        self._s3Client.putStream(part, destPath, headers, function (err, s3Response) {
          if (err) {
            res.writeHead(500, {'content-type': 'application/json'});
            res.end(JSON.stringify({ statusCode: 500, error: err.toString() }));
            return;
          }

          res.writeHead(res.statusCode, {'content-type': 'application/json'});
          if (s3Response.statusCode === 200) {
            res.end(JSON.stringify({ statusCode: 200, uri: prefix + '/raw' + destPath}));
          }
          else {
            var bufs = [];
            s3Response.on('data', function (d){ bufs.push(d); });
            s3Response.on('end', function () {
              // just return the ugly xml body for now
              var body =Buffer.concat(bufs).toString();
              res.end(JSON.stringify({ statusCode: s3Response.statusCode, error: body }));
            });
          }
        });
      });

      form.on('close', onUnexpectedEnd);
      form.parse(req);
    };
  };

  this.serveImage = function (opts) {
    if (!opts) opts = {};
    opts.prefix = opts.prefix || "";
    opts.cacheMaxAge = opts.cacheMaxAge || 220752000; // 1year, 60*60*24*7*365


    return function (req, res) {
      var url = req.url.substring(opts.prefix.length);
      //20-20 min
      //20x20 resize, center crop
      //20+20 max
      var parts = url.split('/');
      var sizeOptions = parts[1];
      var path = '/' + parts.slice(2).join('/');

      self._s3Client.get(path).on('response', function(s3res){
        if (opts.cacheMaxAge) {
          res.setHeader('Cache-Control: max-age', opts.cacheMaxAge); // 1 week
        }

        s3res.on('error', function (err) {
          self._handleError(500, res, err);
        });

        if (sizeOptions === 'raw') {
          return s3res.pipe(res);
        }

        if (s3res.statusCode !== 200) {
          return self._handleError(s3res.statusCode, res, new Error('not found'));
        }

        self._transform(gm(s3res), sizeOptions)
          .stream()
          .pipe(res);
      })
      .on('error', function (err) {
          self._handleError(500, res, err);
      })
      .end();
    };
  };

  this.serveFile = function (opts) {
    if (!opts) opts = {};
    opts.prefix = opts.prefix || "";
    opts.cacheMaxAge = opts.cacheMaxAge || 220752000; // 1year, 60*60*24*7*365


    return function (req, res) {
      var url = req.url.substring(opts.prefix.length);

      self._s3Client.get(url).on('response', function(s3res){
        if (opts.cacheMaxAge) {
          res.setHeader('Cache-Control: max-age', opts.cacheMaxAge); // 1 week
        }

        s3res.on('error', function (err) {
          self._handleError(500, res, err);
        });

        return s3res.pipe(res);
      })
      .on('error', function (err) {
          self._handleError(500, res, err);
      })
      .end();

    };
  };

  this._handleError = function (code, res, err) {
    res.writeHead(code, {'content-type': 'application/json'});
    res.end(JSON.stringify({ statusCode: code, error: err.toString() }));
  };

  this._transform = function (obj, optionsString) {
    var size, w, h;
    if (optionsString.indexOf('x') > -1) {
      size = optionsString.split('x'); w = size[0]; h = size[1];
      return obj.resize(w, h, '^').gravity('Center').crop(w, h, 0, 0);
    }
    else if (optionsString.indexOf('+') > -1) {
      size = optionsString.split('+'); w = size[0]; h = size[1];
      return obj.resize(w, h, '^');
    }
    else if (optionsString.indexOf('-') > -1) {
      size = optionsString.split('-'); w = size[0]; h = size[1];
      return obj.resize(w, h);
    }
    else {
      return obj;
    }
  };
};

example/server.js
var http = require('http')
  , Framer = require('../index')
  , PORT = process.env.PORT || 3002
  , AWS_KEY = process.env.AWS_KEY || '<yours here>'
  , AWS_SECRET = process.env.AWS_SECRET || '<yours here>'
  , AWS_S3BUCKET = process.env.AWS_S3BUCKET || '<yours here>';


var framer = new Framer({
  s3: {
    secure: false,
    key: AWS_KEY,
    secret: AWS_SECRET,
    bucket: AWS_S3BUCKET
  }
});

// authHandler is a function that gets passed the value of the authorization field
// and if valid, passes along an optional value to use as a path prefix when storing
// files. This is helpful for authorization obviously and in identifying ownership of a file.
var handleUpload = framer.handleUpload({
  authHandler: function (value, cb) {
    cb(null, value);
  }
});

var serveImage = framer.serveImage({ prefix: '/img', cacheMaxAge: 3600 });
var serveFile = framer.serveFile({ prefix: '/file', cacheMaxAge: 3600 });

var server = http.createServer(function(req, res) {
  if (req.url === '/') {
    res.writeHead(200, {'content-type': 'text/html'});
    res.end(
      '<form action="/upload" enctype="multipart/form-data" method="post">'+
      '<input type="text" name="authorization"><br>'+
      '<input type="file" name="upload"><br>'+
      '<input type="submit" value="Upload">'+
      '</form>'
    );
  } else if (req.url === '/upload' && req.headers['content-type']) {
    handleUpload(req, res);
  }
  else if (req.url.indexOf('/img/') === 0) {
    serveImage(req, res);
  }
  else if (req.url.indexOf('/file/') === 0) {
    serveFile(req, res);
  }
  else {
    res.writeHead(404, {'content-type': 'applicaiton/json'});
    res.end(JSON.stringify({ statusCode: 404 }));
  }
});

server.listen(PORT, function() {
  console.info('listening on http://0.0.0.0:'+PORT+'/');
});

